docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,295 @@
+   Due to its popularlity, there are several existing tools for this, including
+   [wait-for-it](https://github.com/vishnubob/wait-for-it),
+   [dockerize](https://github.com/jwilder/dockerize), and
+   [wait-for](https://github.com/Eficode/wait-for).
+   The TCP mechanism is portable across platforms
+   and will work in distributed multi-node scenarios.
+
+At least three new ports will be added to EdgeX for sequencing purposes:
+
+* `bootstrap` port. This port will be opened once first-time initialization has been completed.
+* `tokens_ready` port. This port signals that secret-store tokens have been provisioned and are valid.
+* `ready_to_run` port. This port will be opened once stateful services have completed initialization
+  and it is safe for the majority of EdgeX core services to start.
+
+The stateless EdgeX services should block on `ready_to_run` port.
+
+### Docker-specific service changes
+
+#### "As-is" startup flow
+
+The following diagram shows the "as-is" startup flow.
+
+![AS-IS](./bootstrap-as-is.jpg)
+
+There are several components being removed via activity unrelated with this ADR.
+These proposed edits are shown to reduce clutter in the TO-BE diagram.
+* secrets-setup is being eliminated through a separate ADR to eliminate TLS for single-node usage.
+* kong-migrations is being combined with the kong service via an entrypoint script.
+* bootstrap-redis will be incorporated into the Redis entrypoint script to set the Redis password before Redis starts to fix the time delay before a Redis password is set.
+
+#### "To-be" startup flow
+
+The following diagram shows the "to-be" startup flow.
+Note that the bootstrap flows are always processed,
+but can be short-circuited.
+
+![TO-BE](./bootstrap-to-be.jpg)
+
+Another difference to note in the "to-be" diagram is that
+the Vault depdendency on Consul reversed
+[in order to provide better security](https://github.com/edgexfoundry/edgex-docs/issues/296).
+
+#### New Bootstrap/RTR container
+
+The purpose of this new container is to:
+
+* Inject entrypoint scripts into third-party containers
+  (such as Vault, Redis, Consul, PostgreSQL, Kong)
+  in order to perform first-time initialization and
+  wait on service dependencies
+* Raise the `bootstrap` semaphore
+* Wait on dependent semaphores required to raise the `ready_to_run` semaphore
+  (these are the stateful components such as databases,
+  and blocking waiting for sercret store tokens to be provisioned)
+* Raise the `ready_to_run` semaphore
+* Wait forever (in order to leave TCP sockets open)
+
+## Consequences
+
+### Benefits
+
+This ADR is expected to yield the following benefits after completion of the related engineering tasks:
+
+* Standardization of the stage-gate mechanism.
+* Standardized approach to component initialization in Docker.
+* Reduced fragility in the framework startup flow.
+* Vault no longer uses Consul as its data store.
```

Comment:
and uses what instead, the filesystem right?

Reply:
Yes, file system.